pkg/sql/schemachanger/scplan/internal/scstage/build.go

```diff
@@ -474,27 +475,74 @@ func (bc buildContext) computeExtraJobOps(s, next *Stage) []scop.Op {
 			return nil
 		}
 		var ops []scop.Op
+		var newlyUnreferenced []catid.DescID
+		allSet, newlyUnreferencedSet, unreferenced := computeUnreferencedDescriptors(
+			bc.targetState.Targets, s,
+		)
+		if !unreferenced.Empty() && revertible {
+			panic(errors.AssertionFailedf("cannot remove references to descriptors " +
+				"while change is revertible"))
+		}
 		if next == nil {
 			// The terminal mutation stage needs to remove references to the schema
 			// changer job in the affected descriptors.
-			ops = bc.removeJobReferenceOps()
+			ops = bc.removeJobReferenceOps(allSet.Union(newlyUnreferencedSet))
 		} else {
+			// We only need to set newlyUnreferenced if this is not the last stage of
+			// job.
+			newlyUnreferenced = newlyUnreferencedSet.Ordered()
 			const initialize = false
-			ops = bc.setJobStateOnDescriptorOps(initialize, revertible, s.After)
+			ops = append(ops,
+				bc.setJobStateOnDescriptorOps(initialize, revertible, s.After, unreferenced)...)
+			ops = append(ops,
+				bc.removeJobReferenceOps(newlyUnreferencedSet)...)
```

Comment:
I didn't fully follow this block of code and it seems essential. 

IIUC, `newlyUnreferencedSet` is the set of descriptor IDs that have transitioned to a terminal absent status after executing stage `s`. 

1. if `next == nil`, we want to remove job reference from all "existing" descriptors, shouldn't it then be `allSet.Minus(unreferenced)`. I didn't quite understand the `allSet.Union(newlyUnreferencedSet)` here.
2. `else`, we will update job status in all descriptors excluding `unreferenced` and further remove job reference from all `newlyUnreferencedSet`. This part of code made sense to me.

Reply:
Right, it was supposed to be `allSet.Difference(unreferenced).Union(newlyUnreferencedSet)` and then I got distracted mid-refactor. We need to remove the state from everything that wasn't previously unreferenced. Thanks for catching it